### PR TITLE
doc: Adds documents to DomainMatrix class

### PR DESCRIPTION
**Pull request by hyadav2k**
Reference to Task1 for sympy#20987

Collaborative PR with @ks147

#### Brief description of what is fixed or changed
Adds docs for DomainMatrix with the explanation of its methods and usage.

#### RELEASE NOTES
NO ENTRY

### CODE REVIEW

**Comment by hyadav2k** [comment]
@ks147, I messed up the previous PR, while resolving conflicts. But it has now been fixed, I have resolved all conflicts in this PR.
You can review them.
Thanks.

File: sympy/polys/matrices/domainmatrix.py [hunk 1]
```diff
@@ -36,14 +100,13 @@ def from_rep(cls, ddm):
 
     @classmethod
     def from_list_sympy(cls, nrows, ncols, rows, **kwargs):
```
Comment (hyadav2k) [inline sympy/polys/matrices/domainmatrix.py:102]: You should add other sections like in the other methods like `Parameters`, `Returns` and `See Also` if required.
The same can be done for `from_Matrix`  and `eye`.

Comment (ks147) [reply sympy/polys/matrices/domainmatrix.py:102]: Ok.

File: sympy/polys/matrices/domainmatrix.py [hunk 2]
```diff
@@ -100,62 +150,100 @@ def from_Matrix(cls, M, **kwargs):
...
+        >>> domainmatrix1 = DomainMatrix([
+        ...    [ZZ(1), ZZ(2)],
+        ...    [ZZ(3), ZZ(4)]], (2, 2), ZZ)
+        >>> domainmatrix2 = domainmatrix1.convert_to(QQ)
+
+        >>> dM1_new, dM2_new = domainmatrix1.unify(domainmatrix2)
+        >>> dM1_new
+        DomainMatrix([[1, 2], [3, 4]], (2, 2), QQ)
+        >>> dM2_new
+        DomainMatrix([[1, 2], [3, 4]], (2, 2), QQ)
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: @hyadav2k I feel like you could let my example remain. It seems to be written in a more simpler way.

File: sympy/polys/matrices/domainmatrix.py [hunk 3]
```diff
@@ -170,18 +258,33 @@ def unify(self, other):
...
+        >>> domainmatrix = DomainMatrix([
+        ...    [ZZ(1), ZZ(2)],
+        ...    [ZZ(3), ZZ(4)]], (2, 2), ZZ)
+
+        >>> domainmatrix.to_Matrix()
         Matrix([
-        [1, 2],
-        [2, 3]])
+            [1, 2],
+            [3, 4]])
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: I think variable name A is simpler so just let this example also remain as it is.

File: sympy/polys/matrices/domainmatrix.py [hunk 4]
```diff
@@ -196,19 +299,39 @@ def __repr__(self):
...
-        >>> A = DomainMatrix([[ZZ(1), ZZ(2), ZZ(3)]], (1, 3), ZZ)
-        >>> B = DomainMatrix([[QQ(-1, 2), QQ(1, 2), QQ(1, 3)]],(1, 3), QQ)
+        >>> domainmatrix1 = DomainMatrix([
+        ...    [ZZ(1), ZZ(2), ZZ(3)]], (1, 3), ZZ)
+        >>> domainmatrix2 = DomainMatrix([
+        ...    [QQ(-1, 2), QQ(1, 2), QQ(1, 3)]],(1, 3), QQ)
 
-        >>> A.hstack(B)
+        >>> domainmatrix1.hstack(domainmatrix2)
         DomainMatrix([[1, 2, 3, -1/2, 1/2, 1/3]], (1, 6), QQ)
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: Let the example written by me remain as it is.

File: sympy/polys/matrices/domainmatrix.py [hunk 5]
```diff
@@ -303,71 +655,114 @@ def pow(A, n):
...
-        DomainMatrix([[1, 1]], (1, 2), QQ)
+        >>> domainmatrix = DomainMatrix([
+        ...    [ZZ(1), ZZ(1)],
+        ...    [ZZ(0), ZZ(1)]], (2, 2), ZZ)
+        >>> dMF = domainmatrix.convert_to(QQ)
+        >>> dMF
+        DomainMatrix([[1, 1], [0, 1]], (2, 2), QQ)
+
+        >>> dMF.nullspace()
+        DomainMatrix([], (0, 2), QQ)
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: Let the original example remain as it is. Here also

Comment (ks147) [reply sympy/polys/matrices/domainmatrix.py]: @hyadav2k. The reason I am saying let the examples remain is that you've generally used variable names like dMF, domainmatrix1 etc. and I feel A, B as written by me is a better choice. Obviously, this is a very small issue but I still we should stick to A & B

File: sympy/polys/matrices/domainmatrix.py [hunk 6]
```diff
@@ -303,71 +655,114 @@ def pow(A, n):
...
-        DomainMatrix([[2/3, 1/3, 1/6], [1/3, 2/3, 1/3], [0, 0, 1/2]], (3, 3), QQ)
+        >>> domainmatrix = DomainMatrix([
+        ...    [ZZ(1), ZZ(1)],
+        ...    [ZZ(0), ZZ(1)]], (2, 2), ZZ)
+        >>> dMF = domainmatrix.convert_to(QQ)
+        >>> dMF
+        DomainMatrix([[1, 1], [0, 1]], (2, 2), QQ)
+
+        >>> dMF.inv()
+        DomainMatrix([[1, -1], [0, 1]], (2, 2), QQ)
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: Change it back to the original example.

File: sympy/polys/matrices/domainmatrix.py [hunk 7]
```diff
@@ -433,22 +852,48 @@ def lu(self):
...
+        >>> from sympy.polys.matrices import DomainMatrix
+        >>> domainmatrix1 = DomainMatrix([
+        ...    [QQ(1), QQ(2)],
+        ...    [QQ(3), QQ(4)]], (2, 2), QQ)
+        >>> domainmatrix2 = DomainMatrix([
+        ...    [QQ(1), QQ(1)],
+        ...    [QQ(0), QQ(1)]], (2, 2), QQ)
+
+        >>> domainmatrix1.lu_solve(domainmatrix2)
+        DomainMatrix([[-2, -1], [3/2, 1]], (2, 2), QQ)
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: Example

File: sympy/polys/matrices/domainmatrix.py [hunk 8]
```diff
@@ -493,9 +953,46 @@ def eye(cls, n, domain):
...
+        ...    [ZZ(1), ZZ(2)],
+        ...    [ZZ(3), ZZ(4)]], (2, 2), ZZ)
+        >>> domainmatrix2 = DomainMatrix([
+        ...    [ZZ(1), ZZ(1)],
+        ...    [ZZ(0), ZZ(1)]], (2, 2), ZZ)
+
+        >>> domainmatrix1.__eq__(domainmatrix1)
+        True
+        >>> domainmatrix1.__eq__(domainmatrix2)
+        False
```
Comment (ks147) [inline sympy/polys/matrices/domainmatrix.py]: Rename the variables A & B

**Review by ks147 — COMMENTED** [review]
Mostly it's just the variable names that need to be changed. Rest all looks good to me, you've added a Raises and Parameter section to most of the functions (which is good).

File: sympy/polys/matrices/domainmatrix.py (hunk 2 above)
Comment (hyadav2k) [reply sympy/polys/matrices/domainmatrix.py]: Sorry, will use your example here, it looks better.

**Comment by hyadav2k** [comment]
@ks147, even though there is no logical explanation as to why we should use common names like `A` and `B` over `domainmatrix` etc, but it seems fine with me. It will shorten and simplify the docstring, so I will change them.

The reason I am not using most of your examples is either they have a wrong indentation or they are complex in terms of size, and in docstring, we are supposed to have examples that can also be easily solved using mental arithmetic.

File: sympy/polys/matrices/domainmatrix.py (hunk 3 above)
Comment (hyadav2k) [reply sympy/polys/matrices/domainmatrix.py]: Will do.

File: sympy/polys/matrices/domainmatrix.py (hunk 8 above)
Comment (hyadav2k) [reply sympy/polys/matrices/domainmatrix.py]: Sure.

**Comment by ks147** [comment]
> @ks147, even though there is no logical explanation as to why we should use common names like `A` and `B` over `domainmatrix` etc, but it seems fine with me. It will shorten and simplify the docstring, so I will change them.
> 
> The reason I am not using most of your examples is either they have a wrong indentation or they are complex in terms of size, and in docstring, we are supposed to have examples that can also be easily solved using mental arithmetic.

I didn't notice the wrong indentation, my bad. I suppose my examples could have been better. And yes, there's no logical explanation for using A, B over domainmatrix , but I think you should still change it just for aesthetic reasons. Just make these changes then I'll merge your pull request.

**Comment by hyadav2k** [comment]
@ks147, I have replaced the names and the examples you provided.
As I left a review above, you need to add other sections in the methods you added, before you merge.
(Don't merge it directly after, first we need to get the PR reviewed by the mentor.)

Kindly review the latest commits.
Thanks.

**Comment by ks147** [comment]
@hyadav2k, I just resolved a merge commit with the master and after that it's showing some trailing whitespace and failing code quality test. I have tried to remove the 'supposed' whitespace but so far I have been unsuccessful. Just remove the whitespace and then request a review of the PR so I can merge.

**Comment by hyadav2k** [comment]
@ks147, you should refer to https://github.com/sympy/sympy/wiki/Development-workflow#be-sure-that-all-tests-of-sympy-pass, use `./bin/strip_whitespace <file>` to fix whitespaces.

Don't merge the PR into the branch for now.

@oscarbenjamin, Kindly review this PR for docs in `DomainMatrix`.
I was thinking we should merge this PR, and we can open a new PR for the `rst `file on which I and @ks147 or anyone more willing can work. The main reason to make a new PR for the `rst` file is that so that the basic docs can be added to master and if someone wants to get started with working on DomainMatrix, they will have something to look at, else if we will go for `rst `file in this PR, it will take time.

Kindy share your opinion.
Thanks.

**Comment by oscarbenjamin** [comment]
This PR is between your two branches. If you want this to be merged into sympy then it first needs to be merged into a single branch.

**Comment by ks147** [comment]
@hyadav2k , I think we need to first merge your branch into mine. Then after that @oscarbenjamin can review it

**Comment by hyadav2k** [comment]
@ks147, Sure once all the tests pass, kindly merge.
Thanks.

**Pull request merged**